### PR TITLE
Program-Couse-Tags (Many-to-Many)-to-Many

### PR DESCRIPTION
**Pull request by chazkiker2**
# What's here (& why)

Add new 'Tag' entity and attach a Many-to-One-to-Many relationship between `Program`-to-`ProgramTags`-to-`Tag`.

- `Program` entities have a many-to-one-to-many relationship with `Tag` entities by way of a joined-entity `ProgramTags`. 
- `Course` entities have a many to many relationship with `ProgramTags` 
- so Many `Courses` can have Many `ProgramTags`... and `ProgramTags` have Many `Programs` and Many `Tags`... so this is a complex relationship between the three entities `Program`, `Tag`, and `Course`. It's a (Many-to-one-to-Many)-to-Many relationship.
- Adding any `Tag` to any `Course` is explicitly and strongly tied to whether that `Tag` is part of the `Program` which has ownership over this `Course`

In the future, I will be adding more endpoints and operations for how to interact with that complex relationship, but for now it's in a very strong place.

# Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# Change Status

- [x] Complete, tested, ready to review and merge
- [ ] Complete, but not tested (may need new tests)
- [ ] Incomplete / work-in-progress, PR is for discussion or feedback
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes 
- [x] There are no merge conflicts

### CODE REVIEW

File: src/main/java/com/lambdaschool/oktafoundation/services/HelperFunctions.java [hunk 1]
```diff
@@ -37,4 +37,42 @@
 
 	User getCallingUser();
 
+	// THIS may be a good/simple/customizable solution to adding
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/HelperFunctions.java:40]: ALL OF THIS CODE IS MEANT TO BE COMMENTED OUT

This is an interface. So these are all simply Method Signatures for some REAL CLASS to implement. 

These are not in the scope of this PR, so I didn't want to add these changes in here... but I also didn't want to forget this as a possibility for future changes

File: src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+	@Override
+	public Tag get(long tagid) {
+		return tagRepository.findById(tagid)
+				.orElseThrow(() -> new TagNotFoundException(tagid));
+	}
+
+	@Override
+	public Tag get(String title) {
+		return null;
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java:48]: null on purpose

File: src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	@Override
+	public List<Tag> getByProgram(long programid) {
+		return null;
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java:53]: null on purpose

File: src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	@Override
+	public List<Tag> getByProgram(String programname) {
+		return null;
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java:58]: null on purpose

File: src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	@Override
+	public List<Tag> getByCourse(long courseid) {
+		return null;
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java:63]: null on purpose

File: src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	@Override
+	public List<Tag> getByCourse(String coursename) {
+		return null;
```
Comment (chazkiker2) [inline src/main/java/com/lambdaschool/oktafoundation/services/TagServiceImpl.java:68]: null on purpose

**Review by chazkiker2 — COMMENTED** [review]
Ready for Merge

File: src/main/java/com/lambdaschool/oktafoundation/services/HelperFunctions.java (hunk 1 above)
Comment (chazkiker2) [reply src/main/java/com/lambdaschool/oktafoundation/services/HelperFunctions.java:40]: In fact, these additions may not even be in-scope for this month. But would still love to leave this in for some future contributor.

**Review by dav1dchang — APPROVED** [review]
Great job!

**Review by simonesquad — APPROVED** [review]
Approved thanks.

**Pull request merged**